run-tests.php

```diff
@@ -3716,6 +3712,80 @@ private function mergeSuites(array &$dest, array $source): void
     }
 }
 
+class SkipCache
+{
+    private bool $keepFile;
+
+    private array $skips = [];
+    private array $extensions = [];
+
+    private int $hits = 0;
+    private int $misses = 0;
+    private int $extHits = 0;
+    private int $extMisses = 0;
+
+    public function __construct(bool $keepFile)
+    {
+        $this->keepFile = $keepFile;
+    }
+
+    public function checkSkip(string $php, string $code, string $checkFile, string $tempFile, array $env): string
+    {
+        $key = $php;
+
+        // HACK: extension tests frequently use something like <?php require 'skipif.inc';
+        // for skip checks. These have to be cached separately.
+        if (preg_match('/require|include|__FILE__|__LINE__/', $code)) {
+            $dir = dirname($checkFile);
+            $key = "$php => $dir";
+        }
+
+        if (isset($this->skips[$key][$code])) {
+            $this->hits++;
+            return $this->skips[$key][$code];
```

Comment:
If keepFile is enabled and we have a cache hit, then we'll not keep the file (by dint of not creating it in the first place).

Might make sense to simply not cache if keepFile is enabled?

Reply:
Just made it save the file on early exit.